src/main/java/com/surrealdb/connection/SurrealWebSocketConnection.java

```diff
@@ -0,0 +1,148 @@
+package com.surrealdb.connection;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.surrealdb.connection.exception.*;
+import com.surrealdb.connection.model.RpcRequest;
+import com.surrealdb.connection.model.RpcResponse;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.java_websocket.client.WebSocketClient;
+import org.java_websocket.exceptions.WebsocketNotConnectedException;
+import org.java_websocket.handshake.ServerHandshake;
+
+import java.lang.reflect.Type;
+import java.net.ConnectException;
+import java.net.NoRouteToHostException;
+import java.net.URI;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * @author Khalid Alharisi
+ */
+@Slf4j
+public class SurrealWebSocketConnection extends WebSocketClient implements SurrealConnection {
+    private final Gson gson;
+    private final Map<String, CompletableFuture<?>> callbacks;
+    private final Map<String, Type> resultTypes;
+
+    // precomputed private variables
+    private final Pattern RECORD_ALREADY_EXITS_PATTERN = Pattern.compile("There was a problem with the database: Database record `(.+):(.+)` already exists");
+
+    @SneakyThrows
+    public SurrealWebSocketConnection(String host, int port){
+        super(URI.create("ws://"+host+":"+port+"/rpc"));
```

Comment:
This might be `wss://` in production environments, so it might be better to allow passing the full url?

Reply:
I added a boolean parameter `useTls` that controls the protocol used. I think these params (hostname, port, useTLS?) are simpler for the user since the user might use an unsupported protocol like `http` or forget to append the `/rpc` path.